src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -354,6 +431,8 @@ void MarkCustomAttributes (ICustomAttributeProvider provider)
 			if (!provider.HasCustomAttributes)
 				return;
 
+			bool markOnUse = _context.KeepUsedAttributeTypesOnly && Annotations.GetAction (GetAssemblyFromCustomAttributeProvider (provider)) == AssemblyAction.Link;
```

Comment:
I thought about handling the marking of CustomAttributes this way, i.e. down in   `MarkCustomAttributes` and by checking the assembly action.  I decided against this route for a few reasons.

1) I felt it was much more opaque.  If there were ever a bug again in marking everything in an assembly I'd easily stumble on `MarkEntireType` (or in my PR I had a method with the same purpose).  I'd then see that that method didn't do anything with custom attributes and I thought that could easily lead to a red herring situation.

2) Accessing the Assembly of every `ICustomAttributeProvider` object we call `MarkCustomAttributes` and then looking up that assembly in a dictionary seems much more costly than just looping back in `MarkEntireType` and calling `MarkCustomAttribute`.

3) From a tracing perspective, I think it is more accurate to mark the custom attributes back in `MarkEntireType`.  The single biggest reason the custom attribute is being marked is because the assembly has the copy action.  I'm not saying the tracing is currently setup correctly to draw a clear connection back to the Assemblies Copy action, but if we ever wanted to ensure this scenario was traced in a sensible way, I think it would be easier to wire that up if we marked the CustomAttributes directly from` MarkEntireType`

I'm not necessarily opposed to how you did it.  But I did think it was worth mentioning why I explicitly did not pick this approach.

Reply:
I didn't want to copy the CA marking logic. We are heading down the path where linker has deeper knowledge about framework attributes and duplicating that for copy and link actions is something I want to avoid.